src/modules/fetchPage.ts

```diff
@@ -1,54 +1,81 @@
 
 export interface PageData {
 	url: string;
 	html: string;
 }
 
+export interface FetchOptions extends RequestInit {
+	method?: 'GET' | 'POST';
+	body?: string | FormData | URLSearchParams;
+	headers?: Record<string, string>;
+}
+
+export class FetchError extends Error {
+	url: string;
+	status: number;
+	constructor(message: string, details: { url: string; status: number }) {
+		super(message);
+		this.name = 'FetchError';
+		this.url = details.url;
+		this.status = details.status;
+	}
+}
+
+/**
+ * Fetch a page from the server, return it and cache it.
+ */
+
 export async function fetchPage(
 	this: Swup,
-	url: string,
-	data: PageLoadOptions = {}
+	url: URL | string,
+	options: FetchOptions & { triggerHooks?: boolean } = {}
 ): Promise<PageData> {
-	const headers = this.options.requestHeaders;
-	const { url: requestURL } = Location.fromUrl(url);
+	const { url: requestUrl } = Location.fromUrl(url);
+
+	if (this.cache.has(requestUrl)) {
+		const page = this.cache.get(requestUrl) as PageData;
+		if (options.triggerHooks !== false) {
+			await this.hooks.trigger('pageLoaded', { page, cache: true });
+		}
+		return page;
+	}
 
-	const cachedPage = this.cache.get(requestURL);
-	if (cachedPage) {
-		await this.hooks.trigger('pageLoadedFromCache', { page: cachedPage });
-		return Promise.resolve(cachedPage);
+	const headers = { ...this.options.requestHeaders, ...options.headers };
+	options = { ...options, headers };
+
+	// Allow hooking before this and returning a custom response-like object (e.g. custom fetch implementation)
+	const response = await this.hooks.trigger(
+		'fetchPage',
+		{ url: requestUrl, options },
+		async (context, { url, options, response }) => await (response || fetch(url, options))
+	);
+
+	const { status, url: responseUrl } = response;
+	const html = await response.text();
+
+	if (status === 500) {
+		this.hooks.trigger('serverError', { status, response, url: responseUrl });
+		throw new FetchError(`Server error: ${responseUrl}`, { status, url: responseUrl });
+	}
+
+	if (!html) {
+		throw new FetchError(`Empty response: ${responseUrl}`, { status, url: responseUrl });
 	}
 
-	const page = await new Promise<PageData>((resolve, reject) => {
-		fetch(url, { ...data, headers }, (request) => {
-			const { status, responseText, responseURL } = request;
-			const { url } = Location.fromUrl(responseURL || requestURL);
-
-			if (status === 500) {
-				this.hooks.trigger('serverError', { url, status, request });
-				reject(requestURL);
-				return;
-			}
-
-			if (!responseText) {
-				reject(requestURL);
-				return;
-			}
-
-			const html = responseText;
-			const page: PageData = { url, html };
-
-			// Only save cache entry for non-redirects
-			if (requestURL === url) {
-				this.cache.set(url, page);
-			}
-
-			this.hooks.trigger('pageLoaded', { page });
-			resolve(page);
-		});
-	});
+	// Resolve real url after potential redirect
+	const { url: finalUrl } = new Location(responseUrl);
```

Comment:
I keep being irritated by the naming of `Location`. I always think it's the `window.location` or something, even if that of course doesn't make sense. Anyways, a bit late to rename it now, isn't it? 🐌 `SwupLocation` or `SwupURL` would have been more clear IMHO.

Reply:
I think `Location` is perfect. While `window.location` is the very specific `Location` the window is currently at, our `Location`s are free-floating ones that aren't attached to anything at the moment, just pointing at things.

For what it's worth, turbo calls this a `Locatable` which is slightly more weird but kind of the same thing.